x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/LdapRealm.java

```diff
@@ -261,10 +261,13 @@ private static void lookupUserFromSession(
                 metadata.put("ldap_groups", ldapData.groups);
                 metadata.putAll(ldapData.metadata);
```

Comment:
This `metadata` is used in role mappings, and it will now by-default include the `CN` an `mail` user attributes by virtue of the default values for the new settings `user_full_name_attribute` and `user_email_attribute`.

I think this represents a tiny problem when upgrading deployments.
It's possible that some deployments have multiple LDAP/AD realms configured and that they populate the user metadata differently (e.g. some populate it with `CN` others with `mail`, or any other attribute or combination of it). That metadata is probably used in [role mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-roles.html) - there's no point to configure that otherwise.
And so, here's the problem, if all realms suddenly start showing the `CN` and `email` in metadata, it can mess up the role mapping expectation about which users from which realms contain which LDAP attributes.

Bottom line is that I think it's safest to keep populating the user's `metadata` field here, only with the attributes listed in the realm's `metadata` setting (even though, internally, the `LdapMetadataResolver` is used to resolve all attributes together).
Alternatively we could opt to default the `user_full_name_attribute` and `user_email_attribute` settings to null/empty , which would not impact upgrades at all. But I generally prefer the settings with sane default values, so this latter option is less preferred IMO.

Reply:
> I think it's safest to keep populating the user's metadata field here, only with the attributes listed in the realm's metadata setting (even though, internally, the LdapMetadataResolver is used to resolve all attributes together).

Thanks for this great insight. I've updated the code to only populate metadata with the realms configured attributes.